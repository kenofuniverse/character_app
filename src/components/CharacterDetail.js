import React, { useEffect } from 'react'
import { Redirect } from 'react-router'
import PropTypes from 'prop-types'
import { Container } from 'react-bootstrap'
import Loader from 'react-loader-spinner'
import { Row, Col, Image, Table, Button } from 'react-bootstrap'
import moment from 'moment'
import _ from 'lodash'

const CharacterDetail = ({
  detail,
  match,
  history,
  isFetching,
  loadDetail
}) => {
  if (!match.params.id) {
    return <Redirect to="/" />
  }

  useEffect(() => {
    if (match.params.id) {
      loadDetail(match.params.id)
    }
  }, [])

  return (
    <div className="character-detail">
      {isFetching && (
        <Container className="text-center">
          <Loader
            className="mt-10"
            type="Oval"
            color="#007bff"
            height="50"
            width="50"
          />
        </Container>
      )}
      {!_.isEmpty(detail) && !isFetching && (
        <Row>
          <Col sm={12} className="mb-4">
            <Button variant="light" onClick={() => history.goBack()}>
              Go Back
            </Button>
          </Col>
          <Col sm={4}>
            <Image src={detail.image} thumbnail />
          </Col>
          <Col sm={8}>
            <h1 className="character-detail__page-title mb-3">{detail.name}</h1>
            <Table striped bordered hover>
              <tbody>
                <tr>
                  <td>Status</td>
                  <td>{detail.status}</td>
                </tr>
                <tr>
                  <td>Species</td>
                  <td>{detail.species}</td>
                </tr>
                <tr>
                  <td>Type</td>
                  <td>{detail.type}</td>
                </tr>
                <tr>
                  <td>Gender</td>
                  <td>{detail.gender}</td>
                </tr>
                <tr>
                  <td>Origin</td>
                  <td>{detail.origin.name}</td>
                </tr>
                <tr>
                  <td>Location</td>
                  <td>{detail.location.name}</td>
                </tr>
                <tr>
                  <td>Created</td>
                  <td>{moment(detail.created).format("MMMM Do YYYY")}</td>
                </tr>
              </tbody>
            </Table>
          </Col>
        </Row>
      )}
    </div>
  )
}

CharacterDetail.defaultProps = {
  detail: {}
}

CharacterDetail.propTypes = {
  detail: PropTypes.object,
  match: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  isFetching: PropTypes.bool.isRequired,
  loadDetail: PropTypes.func.isRequired
}

export default CharacterDetail
