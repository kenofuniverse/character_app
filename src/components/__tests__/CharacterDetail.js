import React from 'react'
import CharacterDetail from '../CharacterDetail'
import renderer from 'react-test-renderer'

jest.mock('moment', () => () => ({format: () => '2017-11-04T18:48:46.250Z'}));

it('renders correctly', () => {
  const detail = {
    "id": 1,
    "name": "Rick Sanchez",
    "status": "Alive",
    "species": "Human",
    "type": "",
    "gender": "Male",
    "origin": {
      "name": "Earth",
      "url": "https://rickandmortyapi.com/api/location/1"
    },
    "location": {
      "name": "Earth",
      "url": "https://rickandmortyapi.com/api/location/20"
    },
    "image": "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
    "episode": [
      "https://rickandmortyapi.com/api/episode/1",
      "https://rickandmortyapi.com/api/episode/2",
      // ...
    ],
    "url": "https://rickandmortyapi.com/api/character/1",
    "created": "2017-11-04T18:48:46.250Z"
  }
  const tree = renderer
    .create(<CharacterDetail
              detail={detail}
              loadDetail={(id) => {/**/}}
              match={{params: {id: 1}}}
              history={{}}
              isFetching={false}
            />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
