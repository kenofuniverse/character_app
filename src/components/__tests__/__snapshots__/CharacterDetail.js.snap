// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="character-detail"
>
  <div
    className="row"
  >
    <div
      className="mb-4 col-sm-12"
    >
      <button
        className="btn btn-light"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Go Back
      </button>
    </div>
    <div
      className="col-sm-4"
    >
      <img
        className="img-thumbnail"
        src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
      />
    </div>
    <div
      className="col-sm-8"
    >
      <h1
        className="character-detail__page-title mb-3"
      >
        Rick Sanchez
      </h1>
      <table
        className="table table-striped table-bordered table-hover"
      >
        <tbody>
          <tr>
            <td>
              Status
            </td>
            <td>
              Alive
            </td>
          </tr>
          <tr>
            <td>
              Species
            </td>
            <td>
              Human
            </td>
          </tr>
          <tr>
            <td>
              Type
            </td>
            <td>
              
            </td>
          </tr>
          <tr>
            <td>
              Gender
            </td>
            <td>
              Male
            </td>
          </tr>
          <tr>
            <td>
              Origin
            </td>
            <td>
              Earth
            </td>
          </tr>
          <tr>
            <td>
              Location
            </td>
            <td>
              Earth
            </td>
          </tr>
          <tr>
            <td>
              Created
            </td>
            <td>
              2017-11-04T18:48:46.250Z
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
